// sass-embedded/dist/lib/src/exception/exception.js
"use strict";
// Copyright 2020 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * An exception thrown by Sass.
 */
class SassException extends Error {
    /**
     * @param message - The error message.
     * @param formatted - The formatted error message. Includes the message, span,
     *                    and trace.
     * @param [span] - The source span associated with the error.
     * @param [trace] - The trace associated with the error.
     */
    constructor(message, formatted, span, trace) {
        var _a;
        super(message);
        this.message = message;
        this.formatted = formatted;
        this.span = span;
        this.trace = trace;
        if (formatted === '')
            this.formatted = `Error: ${message}`;
        if (trace === '')
            this.trace = undefined;
        // Inject the entire Sass error into the JS stack trace.
        this.stack = (_a = this.stack) === null || _a === void 0 ? void 0 : _a.replace(new RegExp(`^Error: ${message}`), this.formatted);
    }
    toString() {
        return this.formatted;
    }
}
exports.SassException = SassException;
